src/ops.rs

```diff
@@ -35,63 +84,136 @@ bi_enum! {
-      Lt => Gt,
-      Gt => Lt,
-      Lte => Gte,
-      Gte => Lte,
-      And => And,
-      Or => Or,
-      Xor => Xor,
-      Shl => ShlS,
-      ShlS => Shl,
-      Shr => ShrS,
-      ShrS => Shr,
+      Self::Add => Self::Add,
+      Self::Sub => Self::SubS,
+      Self::SubS => Self::Sub,
+      Self::Mul => Self::Mul,
+      Self::Div => Self::DivS,
+      Self::DivS => Self::Div,
+      Self::Rem => Self::RemS,
+      Self::RemS => Self::Rem,
+      Self::Eq => Self::Eq,
+      Self::Ne => Self::Ne,
+      Self::Lt => Self::Gt,
+      Self::Gt => Self::Lt,
+      Self::Le => Self::Ge,
+      Self::Ge => Self::Le,
+      Self::And => Self::And,
+      Self::Or => Self::Or,
+      Self::Xor => Self::Xor,
+      Self::Shl => Self::ShlS,
+      Self::ShlS => Self::Shl,
+      Self::Shr => Self::ShrS,
+      Self::ShrS => Self::Shr,
     }
   }
-  #[inline]
-  pub fn op(self, a: u64, b: u64) -> u64 {
-    const U60: u64 = 0xFFF_FFFF_FFFF_FFFF;
+
+  fn op<T: Num>(self, a: T, b: T) -> T {
     match self {
-      Add => a.wrapping_add(b) & U60,
-      Sub => a.wrapping_sub(b) & U60,
-      SubS => b.wrapping_sub(a) & U60,
-      Mul => a.wrapping_mul(b) & U60,
-      Div => a.checked_div(b).unwrap_or(0),
-      DivS => b.checked_div(a).unwrap_or(0),
-      Mod => a.checked_rem(b).unwrap_or(0),
-      ModS => b.checked_rem(a).unwrap_or(0),
-      Eq => (a == b) as u64,
-      Ne => (a != b) as u64,
-      Lt => (a < b) as u64,
-      Gt => (a > b) as u64,
-      Lte => (a <= b) as u64,
-      Gte => (a >= b) as u64,
-      And => a & b,
-      Or => a | b,
-      Xor => a ^ b,
-      Shl => (a << b) & U60,
-      ShlS => (b << a) & U60,
-      Shr => a >> b,
-      ShrS => b >> a,
+      Self::Add => T::add(a, b),
+      Self::Sub => T::sub(a, b),
+      Self::SubS => T::sub(b, a),
+      Self::Mul => T::mul(a, b),
+      Self::Div => T::div(a, b),
+      Self::DivS => T::div(b, a),
+      Self::Rem => T::rem(a, b),
+      Self::RemS => T::rem(b, a),
+      Self::Eq => T::from_bool(a == b),
+      Self::Ne => T::from_bool(a != b),
+      Self::Lt => T::from_bool(a < b),
+      Self::Le => T::from_bool(a <= b),
+      Self::Gt => T::from_bool(a > b),
+      Self::Ge => T::from_bool(a >= b),
+      Self::And => T::and(a, b),
+      Self::Or => T::or(a, b),
+      Self::Xor => T::xor(a, b),
+      Self::Shl => T::shl(a, b),
+      Self::ShlS => T::shl(b, a),
+      Self::Shr => T::shr(a, b),
+      Self::ShrS => T::shr(b, a),
+    }
+  }
+}
+
+bi_enum! {
+  #[repr(u8)]
+  #[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+  pub enum Ty {
+    "u8":  U8  = 0,
+    "u16": U16 = 1,
+    "u32": U32 = 2,
+    "u60": U60 = 3,
+    "i8":  I8  = 4,
+    "i16": I16 = 5,
+    "i32": I32 = 6,
+  }
+}
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Op {
+  pub ty: Ty,
+  pub op: IntOp,
+}
+
+impl From<u16> for Op {
+  fn from(value: u16) -> Self {
+    unsafe { std::mem::transmute(value) }
```

Comment:
this is UB if `value` is not a valid op; can you either make this a `TryFrom` impl, or an `unsafe Op::from_unchecked`